syz-fuzzer/fuzzer.go

```diff
@@ -283,60 +224,73 @@ func main() {
 		runTest(target, manager, *flagName, config.Executor)
 		return
 	}
-
-	needPoll := make(chan struct{}, 1)
-	needPoll <- struct{}{}
-	fuzzer := &Fuzzer{
-		name:                     *flagName,
-		outputType:               outputType,
-		config:                   config,
-		execOpts:                 execOpts,
-		workQueue:                newWorkQueue(*flagProcs, needPoll),
-		needPoll:                 needPoll,
-		manager:                  manager,
-		target:                   target,
-		timeouts:                 timeouts,
-		faultInjectionEnabled:    r.CheckResult.Features[host.FeatureFault].Enabled,
-		comparisonTracingEnabled: r.CheckResult.Features[host.FeatureComparisons].Enabled,
-		corpusHashes:             make(map[hash.Sig]struct{}),
-		checkResult:              r.CheckResult,
-		fetchRawCover:            *flagRawCover,
-		noMutate:                 r.NoMutateCalls,
-		stats:                    make([]uint64, StatCount),
-		// Queue no more than ~3 new inputs / proc.
-		parallelNewInputs: make(chan struct{}, int64(3**flagProcs)),
-		resetAccState:     *flagResetAccState,
-	}
-	gateCallback := fuzzer.useBugFrames(r, *flagProcs)
-	fuzzer.gate = ipc.NewGate(gateSize, gateCallback)
-
-	for needCandidates, more := true, true; more; needCandidates = false {
-		more = fuzzer.poll(needCandidates, nil)
-		// This loop lead to "no output" in qemu emulation, tell manager we are not dead.
-		log.Logf(0, "fetching corpus: %v, signal %v/%v (executing program)",
-			len(fuzzer.corpus), len(fuzzer.corpusSignal), len(fuzzer.maxSignal))
-	}
+	rnd := rand.New(rand.NewSource(time.Now().UnixNano()))
 	calls := make(map[*prog.Syscall]bool)
 	for _, id := range r.CheckResult.EnabledCalls[sandbox] {
 		calls[target.Syscalls[id]] = true
 	}
-	fuzzer.choiceTable = target.BuildChoiceTable(fuzzer.corpus, calls)
-
+	fuzzerObj := fuzzer.NewFuzzer(context.Background(), &fuzzer.Config{
+		Coverage:       config.Flags&ipc.FlagSignal > 0,
+		FaultInjection: r.CheckResult.Features[host.FeatureFault].Enabled,
+		Comparisons:    r.CheckResult.Features[host.FeatureComparisons].Enabled,
+		Collide:        execOpts.Flags&ipc.FlagThreaded > 0,
+		EnabledCalls:   calls,
+		NoMutateCalls:  r.NoMutateCalls,
+		LeakChecking:   r.CheckResult.Features[host.FeatureLeak].Enabled,
+		FetchRawCover:  *flagRawCover,
+		NeedCandidates: make(chan struct{}),
+		MinCandidates:  uint(*flagProcs * 2),
+		NewInputs:      make(chan rpctype.Input),
+	}, rnd, target)
+
+	fuzzerTool := &FuzzerTool{
+		fuzzer:        fuzzerObj,
+		name:          *flagName,
+		outputType:    outputType,
+		manager:       manager,
+		target:        target,
+		timeouts:      timeouts,
+		config:        config,
+		checkResult:   r.CheckResult,
+		resetAccState: *flagResetAccState,
+	}
+	fuzzerObj.Config.Logf = func(level int, msg string, args ...interface{}) {
+		if level == 0 {
+			fuzzerTool.Logf(level, msg, args...)
+		}
+		if level <= 1 {
+			fuzzerTool.logMu.Lock()
```

Comment:
Do we need this mutex? log package should be thread-safe.

Reply:
It's for the serialization of all output -- the same mutex is held while we print a program we are about to execute in `proc.go`.

Reply:
Will something change if we remove it?

Reply:
I think there's a chance of interleaving these log strings with the programs printed in `proc.go`. It will complicate log parsing or might even make it impossible in some cases.